brave/src/main/java/brave/SpanRecorder.java

```diff
@@ -0,0 +1,92 @@
+package brave;
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import zipkin.Annotation;
+import zipkin.BinaryAnnotation;
+import zipkin.Constants;
+import zipkin.Endpoint;
+import zipkin.reporter.Reporter;
+
+/**
+ * This is a mutable object representing a specific span. It must be locked externally.
+ *
+ * <p>Method signatures are based on the zipkin 2 model eventhough it isn't out, yet.
+ */
+final class SpanRecorder {
+  static final Map<String, String> END_ANNOTATION_TO_ADDRESS_TYPE;
+
+  static {
+    Map<String, String> result = new LinkedHashMap<>();
+    result.put(Constants.CLIENT_RECV, Constants.SERVER_ADDR);
+    result.put(Constants.SERVER_SEND, Constants.CLIENT_ADDR);
+    // TODO: add mr/ms -> ma when rfc is finalized
+    END_ANNOTATION_TO_ADDRESS_TYPE = Collections.unmodifiableMap(result);
+  }
+
+  final Endpoint localEndpoint;
+  final Reporter<zipkin.Span> reporter;
+  final zipkin.Span.Builder span;
+  final boolean shared;
+
+  boolean recordTimestampAndDuration;
+  String addressType;
+  Endpoint remoteEndpoint;
+  boolean finished;
+
+  SpanRecorder(Reporter<zipkin.Span> reporter, Endpoint localEndpoint, TraceContext context) {
+    this.localEndpoint = localEndpoint;
+    this.reporter = reporter;
+    span = zipkin.Span.builder();
+    span.traceIdHigh(context.traceId().hi());
+    span.traceId(context.traceId().lo());
+    span.parentId(context.parentId());
+    span.id(context.spanId());
+    span.debug(context.debug());
+    shared = context.shared();
+  }
+
+  /** Annotations that when added, a span should be finished */
+  Set<String> endAnnotations() {
+    return END_ANNOTATION_TO_ADDRESS_TYPE.keySet();
+  }
+
+  void setName(String name) {
+    span.name(name);
+  }
+
+  void addAnnotation(long timestamp, String value) {
+    // don't report server-side timestamp on shared-spans
+    if (shared && value.equals(Constants.SERVER_RECV)) {
+      recordTimestampAndDuration = false;
+    }
+    String maybeAddressType = END_ANNOTATION_TO_ADDRESS_TYPE.get(value);
+    if (maybeAddressType != null) addressType = maybeAddressType;
+    span.addAnnotation(Annotation.create(timestamp, value, localEndpoint));
```

Comment:
@devinsba I refactored everything so there's no intermediate types used. Current impl is more efficient and less code. Thanks for the prod that led to this. Incidentally, the code is more similar to [zipkin-finagle](https://github.com/openzipkin/zipkin-finagle) now.

```
SpanCreationBenchmarks.no_contention:no_contention_elaborateSpan_brave3       thrpt   15  0.572 ± 0.032  ops/us
SpanCreationBenchmarks.no_contention:no_contention_elaborateSpan_brave4       thrpt   15  0.739 ± 0.035  ops/us
SpanCreationBenchmarks.no_contention:no_contention_simpleRootSpan_brave3      thrpt   15  1.393 ± 0.075  ops/us
SpanCreationBenchmarks.no_contention:no_contention_simpleRootSpan_brave4      thrpt   15  2.067 ± 0.097  ops/us
```